src/checkpoint/checkpoint_api.h

```diff
@@ -78,7 +78,7 @@ using SerializedReturnType = std::unique_ptr<SerializedInfo>;
  * \return a \c std::unique_ptr to a \c SerializedInfo containing the buffer
  * with serialized data and the size of the buffer
  */
-template <typename T>
+template <typename... UserTraits, typename T>
```

Comment:
It's a bit weird to me that the template arguments here are in a different order compared to deserialize because of the argument type deduction (while deserialize can't deduce).

@lifflander and @Matthew-Whitlock what do y'all think of potentially making them deducable with something like:

```cpp
template< typename... UserTraits >
struct user_traits {};  // we could even have things like has_trait here
```

But then would have to add a new overload:
```cpp
template <typename T, typename... UserTraits>
SerializedReturnType serialize(T& target, user_traits<UserTraits...>, BufferCallbackType fn = nullptr) 
```

Reply:
Example usage:

```cpp
serialize(my_int, user_traits<shallow>{});
```

Reply:
I swapped the template ordering, but left it without the type deduction for now. Mostly to defer questions about where to place the new argument in the various overloads